### ISSUE TITLE
Question: Overriding downstream version ranges

### ISSUE THREAD

**Issue by acmeijer** [conan-io/conan#5507]
To help us debug your issue please explain:

- [X] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [X] I've specified the Conan version, operating system version and any tool that can be relevant.
- [X] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

Conan version 1.17.0
OS: Ubuntu 18.04

We are currently using a lot of version ranges in all of our conanfiles, typically in the following form e.g. [~1.x], this works well for our purposes.

The problem arises when we want to test a branch of one of our libraries, i.e. one that does not have a version associated with it. An example:

```
appA: 
   - requires libB/[~1.x]
   - requires libC/[~1.x]
```
```
libC:
   - requires libB/[~1.x]
```
Now I would like to test a branch of libB together with appA, so we get:
```
appA:
  - requires libB/<my_branch>
  - requires libC/[~1.x]
```
```
libC:
  - requires libB/[~1.x]
```

This will raise a conflict: 
`ERROR: Version range '~1.x' required by 'libC/1.0.0@mychannel/stable' not valid for downstream requirement 'libB/my_branch@mychannel/stable'`

I am aware that we could use the include_prerelease feature for this but that would require us to specify that in every conanfile (also conanfiles of released libraries) just so that we can test new versions occasionally.

The behaviour if you don't use version ranges is that the top-level conanfile overrides the versions specified in its dependencies. I think that the same behaviour if you do use version ranges would be very useful as well as make the behaviour more consistent across the board.

Does this seem like something that is possible?

**Comment by jgsogo** [comment]
> The behaviour if you don't use version ranges is that the top-level conanfile overrides the versions specified in its dependencies. I think that the same behaviour if you do use version ranges would be very useful as well as make the behaviour more consistent across the board.

I agree with you, I think that overridden dependencies don't need to satisfy any given version range. 

I'm reproducing the issue described and the output is as follows:

```
WARN: libC/1.0@issue/testing requirement libB/[~1.0]@issue/testing overridden by app/1.0@issue/testing to libB/my_branch@issue/testing 
ERROR: Version range '~1.0' required by 'libC/1.0@issue/testing' not valid for downstream requirement 'libB/my_branch@issue/testing'
```

**Comment by jgsogo** [comment]
I've found a problem related to this: we already have [tests checking this behavior](https://github.com/conan-io/conan/blob/df44fea73fa7610a5de3634d7354105a12b69d10/conans/test/unittests/client/graph/version_ranges_graph_test.py#L169), so we need to decide if it is a bug or we need to wait for Conan v2.0. As always, someone could trust in this behavior and consider any change a regression.

**Comment by Minimonium** [comment]
It would be great to have that one for feature testing since they're prerelease versions. :slightly_smiling_face:

**Comment by jgsogo** [comment]
#5713 is merged to incoming `Conan 1.20`, it will be available soon 🎉
